tests/TestCase/Core/FunctionsTest.php

```diff
@@ -91,6 +91,30 @@ public function testDeprecationWarningEnabled(): void
         });
     }
 
+    /**
+     * Test duplicate deprecations emit in cli environment and when run PHPUnit
```

Comment:
Won't emitting duplicates have the drawback of having more noisy output in tests? My thinking was that you might have a method that is called 50+ times and that you don't need a deprecation warning for each time it is called.

Reply:
when we have several testCases each those need to have /catch deprecation warning independently
ّ ّ I had not good Idea to both , catching deprecation with `deprecated()` + not annoying developer

if there is better solution I can revert last commit

Reply:
@markstory  there is a way

having other condition in `deprecationWarning()` method
finding deprecated() method in `$trace` and show all deprecation, otherwise ignoring duplicated warning

if you agree I can do it in next PR

Reply:
> finding deprecated() method in $trace and showing all deprecations, otherwise ignoring duplicated warning

I think that is a reasonable solution. Would we only crawl the stack if the message was identified as a duplicate? Walking the stack could be expensive and create measurable overhead in application code at the beginning of their upgrade process.

Reply:
> Would we only crawl the stack if the message was identified as a duplicate?

yes , just how it is now in this PR

```
        if (
                isset($errors[$checksum]) &&
                (
                    // show all deprecations when phpunit runs
                    //conditions
                )
        )
```